decred/decred/dcr/wire/netaddress.py

```diff
@@ -0,0 +1,222 @@
+def readNetAddress(b, hasStamp):
+    """
+	Reads an encoded NetAddress from b depending on the protocol version and
+	whether or not the timestamp is included per hasStamp.  Some messages like
+	version do not include the timestamp.
+
+	Args:
+		b (ByteArray): The encoded NetAddress.
+		hasStamp (bool): Whether or not the NetAddress has a timestamp.
+
+	Returns:
+		NetAddress: The decoded NetAddress.
+	"""
+    expLen = 30 if hasStamp else 26
+    if len(b) != expLen:
+        raise DecredError(
+            f"readNetAddress wrong length (hasStamp={hasStamp}) expected {expLen}, got {len(b)}"
+        )
+
+    # NOTE: The Decred protocol uses a uint32 for the timestamp so it will
+    # stop working somewhere around 2106.  Also timestamp wasn't added until
+    # protocol version >= NetAddressTimeVersion
+    stamp = b.pop(4).unLittle().int() if hasStamp else 0
+    services = b.pop(8).unLittle().int()
+    ip = b.pop(16)
+    if ip[:12] == ipv4to16prefix:
+        ip = ip[12:]
+
+    # Sigh.  Decred protocol mixes little and big endian.
+    port = b.pop(2).int()
+
+    return NetAddress(ip=ip, port=port, services=services, stamp=stamp,)
+
+
+def writeNetAddress(netAddr, hasStamp):
+    """
+	writeNetAddress serializes a NetAddress depending on the protocol
+	version and whether or not the timestamp is included per hasStamp.  Some
+	messages like version do not include the timestamp.
+
+	Args:
+		netAddr (NetAddress): The peer's NetAddress.
+		hasStamp (bool): Whether to encode the timestamp.
+
+	Returns:
+		ByteArray: The encoded NetAddress.
+	"""
+    # NOTE: The Decred protocol uses a uint32 for the timestamp so it will
+    # stop working somewhere around 2106.  Also timestamp wasn't added until
+    # until protocol version >= NetAddressTimeVersion.
+    b = (
+        ByteArray(netAddr.timestamp, length=4).littleEndian()
+        if hasStamp
+        else ByteArray()
+    )
+
+    # Ensure to always write 16 bytes even if the ip is nil.
+    ip = netAddr.ip
+    if ip is None:
+        ip = ByteArray(length=16)
+    if len(ip) == 4:
+        ip = ipv4to16prefix + ip
+
+    b += ByteArray(netAddr.services, length=8).littleEndian()
+    b += ByteArray(ip, length=16)
+    b += ByteArray(netAddr.port, length=2)
+
+    return b
+
+
+class TCPAddr:
+    """ TCPAddr is a mimic of Go's net.TCPAddr. """
+
+    def __init__(self, ip, port, zone=None):
+        """
+		Args:
+			ip (str or bytes-like): The IP.
+			port (int): The port.
+			zone (str): The IPv6 scoped addressing zone.
+		"""
+        if isinstance(ip, str):
+            ip = decodeStringIP(ip)
+        self.ip = ip
+        self.port = port
+        self.zone = zone
+
+
+def decodeStringIP(ip):
+    """
+	Parse an IP string to bytes.
+
+	Args:
+		ip (str): The string-encoded IP address.
+
+	Returns:
+		bytes-like: The byte-encoded IP address.
+	"""
+    try:
+        return socket.inet_pton(socket.AF_INET, ip)
+    except OSError:
+        return socket.inet_pton(socket.AF_INET6, ip)
```

Comment:
What if this call fails too?

Reply:
Then the exception is raised, which is intended but maybe not great for logging.